### ISSUE TITLE
Lint expressions with a single-return arrow syntax but are multiline (possible update to prefer_expression_function_bodies)

### ISSUE THREAD

**Issue by matuella** [dart-lang/linter#2469]
### Describe the issue
When I turn on the `prefer_expression_function_bodies` rule, it doesn't take into consideration whether my declaration is a multiline return or not, meaning that I can't force the lint to only accept single-line statements if it's a return that uses the arrow syntax.

### Suggested behavior
```dart
// BAD Example 1:
String get veryVerboseGetter => 
    'This is a super verbose String and will have more than 80c';

// GOOD Example 1:
String get veryVerboseGetter {
  return 'This is a super verbose String and will have more than 80c';
}
```

```dart
// BAD Example 2:
final String Function(String, String, String) verboseFunction =
    (param1, param2, param3) => param1 + param2 + param3;

// GOOD Example 2:
final String Function(String, String, String) verboseFunction =
    (param1, param2, param3) {
  return param1 + param2 + param3;
};
```

```dart
// GOOD Example 3:
String get simpleGetter => 'This is not a verbose String';
```

```dart
// GOOD Example 4:
final String Function(String) simpleFunction = (param1) => 'simple' + param1;
```

Because of the fact that we currently can't configure individual rules, I don't know if this should be categorized as a:

1. "modification" on the already existing one, `prefer_expression_function_bodies: Possibly warn that `prefer_expression_function_bodies` should only occur when the return declaration are in the same line; or a
2. new lint, something like `avoid_multiline_expression_function_bodies`: Literally they would only exist to this specific scenario where I am suggesting.

### Rationale
I often find myself dealing with use-cases where I don't want to be forced to use arrow syntax. Such cases are frequent when the body of my expression is correctly linted as a single `return` statement, but this statement is complex enough to justify opening a body, like chaining some calls to a stream/lists or even simply chaining a couple of boolean comparisons.

I know this suggestion might not be trivial and the rationale could sound a little bit off, but I wanted to understand a bit more if this is feasible/acceptable. I can try to contribute to make this happen (if the suggestion of a new/updating existing lint makes sense).

One related issue might be #1850

Thanks in advance!

**Comment by bwilkerson** [comment]
Thanks for opening an issue.

> 1. "modification" on the already existing one ...

The `prefer_expression_function_bodies` lint rule is designed to support the style guide recommendation [CONSIDER using => for simple members.](https://dart.dev/guides/language/effective-dart/usage#consider-using--for-simple-members) The examples in the style guide include multi-line bodies, so I don't think we'd want to change the lint rule to only allow single-line functions.

> 2. new lint, something like `avoid_multiline_expression_function_bodies` ...

We could potentially implement something like that, but I'll note that these two lints would then conflict. Given a method with a single return statement the `prefer_expression_function_bodies` lint would tell you that it should use an expression function body (`=>`), but if you converted it the `avoid_multiline_expression_function_bodies` lint would tell you that it should use a block function body. There wouldn't be any way to satisfy both.

That isn't to say that we won't add such a lint, just that it won't be reasonable to enable both lints, which might not be what you're wanting.

> I often find myself dealing with use-cases where I don't want to be forced to use arrow syntax. Such cases are frequent when the body of my expression is correctly linted as a single `return` statement, but this statement is complex enough to justify opening a body, like chaining some calls to a stream/lists or even simply chaining a couple of boolean comparisons.

I don't know whether either will be appealing, but I can think of a couple of possible work-arounds.

First, you could add an [`ignore` comment](https://dart.dev/guides/language/analysis-options#suppressing-rules-for-a-line-of-code) for the methods that you prefer to not convert. That would remove the lint without requiring you to change your code.

Second, you might in some cases find it useful to reduce the complexity of the expression by introducing local variables. That would have the additional effect that the method body would then consist of more than a single return statement.

**Comment by matuella** [comment]
Yes, I thought about this - that there is a problem with them coexisting -, but the core issue remains not only that they would conflict with each other, but also that the `prefer_expression_function_bodies` shouldn't be changed due a generic personal preference about multiline arrow syntax return statements.

> First, you could add an ignore comment for the methods that you prefer to not convert. That would remove the lint without requiring you to change your code.
> 
> Second, you might in some cases find it useful to reduce the complexity of the expression by introducing local variables. That would have the additional effect that the method body would then consist of more than a single return statement.

1. The ignore workarounds aren't "my kind of solution", because they still "pollute" the file, even though they are a completely sensible solution for a couple of cases;

2. I agree about reducing the complexity of expressions, even though I still think _those_ cases make sense, more likely in situations with very verbose stuff... I will try to start tracking more use cases and come here again with better scenarios, then, maybe, I can try to justify the "why" with a less personal opinion and more effective on the code legibility.

For now, all what you've stated makes complete sense for the linter itself. Thanks for the suggestions and taking your time to answer the issue.

**Comment by matthewkooshad** [comment]
i faced the same blocker on using this lint. all of my Widget build methods looked ugly when i have this lint turned on. i'll have to turn it off to avoid lint comments in my code. not enforcing when multiline would be great, @bwilkerson.

**Comment by bwilkerson** [comment]
@munificent For comments on style issues.

**Comment by munificent** [comment]
Honestly, I would just recommend not using the `prefer_expression_function_bodies` lint at all. I don't think it's good guidance to state that all or even nearly all functions whose body is a single return statement should use `=>`. Lots of times, if the expression is long enough, it's cleaner to give the function a block body.
